File: backend/consultation_packages.py
"""The three packages a Physiotherapy consultation is sold as, and how long each one runs.

The Physiotherapy consultation shelf is not named freely any more. It is one of three
packages, and how long an appointment runs follows from which one was sold rather than from
somebody picking minutes off a row of buttons beside the name. So the package is the input
and the duration is derived from it.

Derived, never accepted from the client, for the same reason price_is_total is: the name on
the row says what was sold and the duration says how much of a physio's day to block out
for it. A client free to send both could store a "Consultation + 20 mins Physio" that books
fifteen minutes, and nothing downstream could tell which of the two was the lie.

Kept in its own module, free of FastAPI and of the database, so the arithmetic that decides
how much of a calendar an appointment takes can be read and tested on its own. Same
arrangement as lead_mapping and branch_routing.

The popup mirrors this table (CONSULTATION_PACKAGES in frontend PackagesBoard.jsx) to draw
the dropdown and preview the length. This side is the authority: it re-derives the duration
on the way in and ignores whatever the form sent for it, so the two can disagree about the
wording without ever disagreeing about the minutes.
"""

from typing import Dict, List, Optional


# Every package contains the consultation itself, and it is the same 45 minutes in each.
CONSULTATION_MINUTES = 45

# The physio that runs on the end of the consultation in the second package.
PHYSIO_ADD_ON_MINUTES = 20


PACKAGES: Dict[str, Dict[str, object]] = {
    "only_consultation": {
        "label": "Only Consultation",
        "minutes": CONSULTATION_MINUTES,
        "includes_session": False,
    },
    "consultation_plus_physio": {
        "label": f"Consultation + {PHYSIO_ADD_ON_MINUTES} mins Physio",
        # One booking, run end to end: the physio follows the consultation in the same
        # slot, so the slot has to be long enough to hold both.
        "minutes": CONSULTATION_MINUTES + PHYSIO_ADD_ON_MINUTES,
        "includes_session": False,
    },
    "consultation_plus_session": {
        "label": "Consultation + 1 Session",
        # Forty-five minutes and not a minute more.
        #
        # The session bundled into this package is booked on its own, whenever the patient
        # is next free; it is not part of this appointment. There is no session length in
        # this system to add to the consultation even if it were -- a session package
        # carries a count, not a duration, and the Create Session form has never asked for
        # one. The 30 sitting on those rows is the schema default, not a length anybody
        # chose.
        #
        # So the appointment is the consultation, and the session is a promise recorded
        # against it. Inventing a number here would block out time in a physio's calendar
        # that nobody agreed to, on every consultation ever sold this way.
        "minutes": CONSULTATION_MINUTES,
        "includes_session": True,
    },
}


def keys() -> List[str]:
    return list(PACKAGES)


def label_of(key: Optional[str]) -> str:
    """What the shelf calls this package, or "" for anything not on it."""
    return str((PACKAGES.get(key or "") or {}).get("label") or "")


def duration_for(key: Optional[str]) -> Optional[int]:
    """How long this package runs, or None where no package was named.

    Raises ValueError on a key this module does not know, rather than falling back to a
    default: an unrecognised package is a form and a server that disagree about what is on
    the shelf, and guessing a length for it would file that disagreement as a bookable
    slot in somebody's day. The caller turns this into a 400.
    """
    if not key:
        return None
    package = PACKAGES.get(key)
    if not package:
        raise ValueError(f"Unknown consultation package: {key}")
    return int(package["minutes"])


def includes_session(key: Optional[str]) -> bool:
    """Whether this package owes the patient a session beyond the appointment itself."""
    return bool((PACKAGES.get(key or "") or {}).get("includes_session"))

File: frontend/src/lib/consultationPackages.js
/**
 * The three packages a Physiotherapy consultation is sold as.
 *
 * Read in two places that must agree: Services & Products, where a package is created and
 * its length previewed, and the Consultation Fee desk, where one is picked and its price
 * collected. Kept here rather than in either screen so neither imports the other — and so
 * there is one list to change when a fourth package appears.
 *
 * Mirrors PACKAGES in backend/consultation_packages.py. `minutes` is the default length the
 * create form pre-fills; Super Admin can edit it per item, and the stored duration_minutes
 * is what gets booked. A test (backend/tests/test_consultation_packages.py)
 * reads this file back and fails if the two lists disagree, because a preview promising 65
 * minutes while the server stores 45 would surface as a physio's afternoon running over
 * rather than as anything on screen.
 */
export const CONSULTATION_PACKAGES = [
  {
    key: "only_consultation",
    label: "Only Consultation",
    minutes: 45,
    breakdown: "45 mins",
    durationLabel: "45 mins",
  },
  {
    key: "consultation_plus_physio",
    label: "Consultation + 20 mins Physio",
    minutes: 65,
    // One booking end to end: the physio follows the consultation in the same slot.
    breakdown: "45 mins + 20 mins = 65 mins",
    durationLabel: "65 mins",
  },
  {
    key: "consultation_plus_session",
    label: "Consultation + 1 Session",
    minutes: 45,
    breakdown: "45 mins + 1 Session",
    // Not "45 mins". A row promising a session that reads as 45 minutes looks like the
    // session was forgotten, wherever it is listed.
    durationLabel: "45 mins + 1 Session",
    // Said out loud in the create form, because "45 mins" against a package whose name
    // promises a session reads like a mistake otherwise. A session has no length in this
    // system to add -- a session package carries a count, not a duration -- and it is
    // booked whenever the patient is next free, so holding time for it here would block
    // out a physio's calendar for an appointment nobody has made.
    note: "The session is booked separately. Only the 45-minute consultation is held in the calendar.",
  },
];

/** The package keys in the order they are read on the shelf, and escalate in. */
export const CONSULTATION_PACKAGE_ORDER = CONSULTATION_PACKAGES.map((p) => p.key);

export const consultationPackageByKey = (key) =>
  CONSULTATION_PACKAGES.find((p) => p.key === key) || null;

/**
 * What one consultation item costs, for the mode the appointment is actually in.
 *
 * Online and offline are priced separately on every catalogue row. The server reads this
 * off the stored item itself when the money is taken — this is the same sum, for showing
 * the figure before anyone presses Collect.
 */
export const consultationItemPrice = (item, mode) =>
  (mode === "online" ? item?.price_online : item?.price_offline) ?? null;
